contracts/TreasuryActions.cdc

```diff
@@ -293,4 +293,82 @@ pub contract TreasuryActionsV4 {
       self.intent = "Update the threshold of signers to ".concat(threshold.toString()).concat(".")
     }
   }
-}
\ No newline at end of file
+
+  pub struct AddSignerUpdateThreshold: MyMultiSigV4.Action {
+    pub let threshold: UInt
+    pub let signer: Address
+    pub let intent: String
+    pub let proposer: Address
+
+    access(account) fun execute(_ params: {String: AnyStruct}) {
+      let treasuryRef: &DAOTreasuryV4.Treasury = params["treasury"]! as! &DAOTreasuryV4.Treasury
+      let manager = treasuryRef.borrowManager()
+
+      // Add Signer
+      manager.addSigner(signer: self.signer)
+      // Update Threshold
+      manager.updateThreshold(newThreshold: self.threshold)
```

Comment:
Looks like the actions are take place async, could be problematic, consider this scenario:
- Treasury starts with one signer
-  Signer proposes to add another signer and increase the threshold to 2
-  The action of adding the signer succeeds
-  The action of updating the threshold fails since the post condition demands signer.length < threshold
-  Error message of failed updating threshold is surfaced to FE when the action is partially failed. 
-  User refreshes the page and see signer length is increased but not the threshold.

Reply:
@linxiao-li  the post condition for `updateThreshold` is checking for `signer.length >= threshold`, which looks correct and i don't see issues there. Did i miss anything here?

Reply:
@amyliu6 base on what I read about the post conditions, it seems like the scope is limited to the function, in this case, line 310 only. At that point, line 308 executed successfully, line 310 fails, the whole transaction is half success half fail, but was treated as 100 percent fail because of the post condition on 310.

Reply:
@linxiao-li the post-condition of `updateThreshold` is asserting that `self.signers.length >= self.threshold` is `true`.  In the scenario you have described, the second signer would be added first, so `self.signers.length` would equal 2.  Then, the threshold is updated to 2, and the post-condition will hold true.

So I think for addSigner/updateThreshold we are fine because you can always add a signer no matter the threshold.

But for removeSigner/updateThreshold, I believe there is an issue.... we should updateThreshold FIRST, then remove the signer after.

If we stick to the correct order of add signer -> update threshold, update threshold -> remove signer, I think we should be fine.  Let me know if you think there is an issue I'm not seeing

Reply:
Got it! I misunderstood the syntax on the post condition, it is saying signer count has to be equal or greater than threshold, otherwise throw an error. That makes a lot more sense! But the issue still exists for the add signer flow, what if the param is add one more signer and increase the threshold to 3? (I know on FE there's logic to forbid such action, but don't trust us :)))) Can we add a check function before? Something like
```
fun canUpdate(newSignerCount:UInt, newThreshold:Uint): Bool {
    return newSignerCount >= newThreshold;
}
if canUpdate {
    manager.updateSigner(signer: self.signer);
    manager.updateThreshold(newThreshold: self.threshold);
}
```

Reply:
Yup exactly! Thanks for pointing it out, though because it made me catch an issue w/ `RemoveSignerUpdateThreshold` (just pushed a fix!) :)

I don't think the issue exists for add signer flow, because the `updateThreshold` post-condition will always cause the entire transaction to fail.  Even if `manager.addSigner` runs successfully, the tx will fail on `manager.updateThreshold` because `self.signers.length >= self.threshold` will not hold true.

Are you saying you want to prevent a user from even proposing such an action? I'm pretty confident that the post-conditions will prevent this scenario from ever occurring (i.e. such an action will not be able to be executed if it violates this rule)

Reply:
> post-condition will always cause the entire transaction to fail. Even if manager.addSigner runs successfully, the tx will fail on manager.updateThreshold because self.signers.length >= self.threshold will not hold true.

Would it roll back the signer amount as well?

Reply:
Just did some research, looks like when there's an error for the transaction, the incurred changes get reverted. @jacksonConrad Is that correct? If that's the case I'm all good. Sorry for the confusion!

Reply:
Yup, that is correct! No worries!